src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -118,20 +115,22 @@ private object VisitCache(ServiceCallSite callSite, RuntimeResolverContext conte
 
         private object ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
         {
-            Dictionary<ServiceCacheKey, object> resolvedServices = serviceProviderEngine.ResolvedServices;
-            if (!resolvedServices.TryGetValue(callSite.Cache.Key, out object resolved))
+            object resolved;
+            ConcurrentDictionary<ServiceCacheKey, object> resolvedServices = serviceProviderEngine.ResolvedServices;
+            if (resolvedServices.TryGetValue(callSite.Cache.Key, out resolved))
             {
-                resolved = VisitCallSiteMain(callSite, new RuntimeResolverContext
-                {
-                    Scope = serviceProviderEngine,
-                    AcquiredLocks = context.AcquiredLocks | lockType
-                });
-
-                serviceProviderEngine.CaptureDisposable(resolved);
-                resolvedServices.Add(callSite.Cache.Key, resolved);
+                return resolved;
             }
 
-            return resolved;
+            // This method uses lock to synchronize access to the resolvedServices dictionary.
+            // but it does not need to lock here because the lock for resolving a singleton is taken outside of this method.
+            resolved = VisitCallSiteMain(callSite, new RuntimeResolverContext
+            {
+                Scope = serviceProviderEngine,
+                AcquiredLocks = context.AcquiredLocks | lockType
+            });
+            serviceProviderEngine.CaptureDisposable(resolved);
+            return resolvedServices.GetOrAdd(callSite.Cache.Key, resolved);
```

Comment:
This should be TryAdd, not GetOrAdd.

Reply:
ExpressionBuilderResolver and ILEmitBuilderResolver, from what I see they use resolved services as IDictionary and call Add and TryGetValue using lock.

If I wanted to use TryAdd I would wanna keep resolvedService variable here as ConcurrentDictionary.

Reply:
I'll push an update up in a few mins, using Add instead of TryAdd but using lock only on the Add.